src/LinAlg/hiopKrylovSolver.cpp

```diff
@@ -0,0 +1,351 @@
+// Lawrence Livermore National Security, LLC, and shall not be used for advertising or
+// product endorsement purposes.
+
+/* implements the Krylov iterative solver
+* @file hiopKrylovSolver.cpp
+* @ingroup LinearSolvers
+* @author Nai-Yuan Chiang <chiang7@lnnl.gov>, LNNL
+* @author Cosmin G. Petra <petra1@lnnl.gov>, LNNL
+*/
+
+#include "hiopKrylovSolver.hpp"
+
+#include "hiopVector.hpp"
+
+#include "hiopOptions.hpp"
+#include <hiopLinAlgFactory.hpp>
+#include "hiopLinearOperator.hpp"
+
+namespace hiop {
+
+  /*
+  * class hiopKrylovSolver
+  */
+  hiopKrylovSolver::hiopKrylovSolver(hiopNlpFormulation* nlp,
+                                     hiopLinearOperator* A_opr,
+                                     hiopLinearOperator* Mleft_opr,
+                                     hiopLinearOperator* Mright_opr,
+                                     hiopVector* x0)
+    : nlp_{nlp},
+      A_opr_{A_opr}, 
+      ML_opr_{Mleft_opr},
+      MR_opr_{Mright_opr},
+      x0_{nullptr}
+  {
+    if(x0) {
+      x0_ = x0->new_copy();
+    }
+  }
+  
+  hiopKrylovSolver::~hiopKrylovSolver()
+  {
+    if(x0_) {
+      delete x0_;
+    }
+  }
+
+  bool hiopKrylovSolver::set_x0(hiopVector& x0)
+  {
+    if(x0_){
+      // just a identity precond
+      x0_->copyFrom(x0);
+    } else {
+      x0_ = x0.new_copy();
+    }
+    return true;
+  }
+
+  /*
+  * class hiopPCGSolver
+  */
+  hiopPCGSolver::hiopPCGSolver(hiopNlpFormulation* nlp,
+                               hiopLinearOperator* A_opr,
+                               hiopLinearOperator* Mleft_opr,
+                               hiopLinearOperator* Mright_opr,
+                               hiopVector* x0)
+    : hiopKrylovSolver(nlp, A_opr, Mleft_opr, Mright_opr, x0),
+      tol_{1e-8},
+      maxit_{100},
+      iter_{0.0},
+      flag_{-1},
+      xk_{nullptr},
+      xmin_{nullptr},
+      res_{nullptr},
+      yk_{nullptr},
+      zk_{nullptr},
+      pk_{nullptr},
+      qk_{nullptr}
+  {
+  }
+
+  hiopPCGSolver::~hiopPCGSolver()
+  {
+    if(xk_) {
+      delete xk_;
+      delete xmin_;
+      delete res_;
+      delete yk_;
+      delete zk_;
+      delete pk_;
+      delete qk_;
+    }
+  }
+
+bool hiopPCGSolver::solve(hiopVector& b)
+{
+  // rhs = 0 --> solution = 0
+  double n2b = b.twonorm();
+  if(n2b == 0.0) {
+    b.setToZero();
+    flag_ = 0;
+    iter_ = 0.;    
+    return true;
+  }
+
+  if(xk_==nullptr) {
+    xk_ = b.alloc_clone();    //iterate x
+    xmin_ = b.alloc_clone();  //iterate which has minimal residual so far
+    res_ = b.alloc_clone();   //minimal residual iterate 
+    yk_ = b.alloc_clone();    //work vectors
+    zk_ = b.alloc_clone();    //work vectors 
+    pk_ = b.alloc_clone();    //work vectors
+    qk_ = b.alloc_clone();    //work vectors
+  }
+
+  //////////////////////////////////////////////////////////////////
+  // Starting procedure
+  //////////////////////////////////////////////////////////////////
+  // starting point
+  if(x0_) {
```

Comment:
I would allocate `x0_` as soon as possible and then, if possible, reuse it for `xk_` to save on one vector